### PR TITLE
EP_RDM/TAGGED implementation for verbs provider.

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
Signed-off-by: evgeny-leksikov <evgeny.leksikov@gmail.com>

Status:
	1) The same amount of passed ep_msg fabtests, no new passed fabtests
	2) Pass rate of mpich tests with Intel(R) MPI Library is 89.5%. Also we have managed to run Intel(R) MPI Benchmarks (IMB-MPI1) over Open MPI.
	
Technical specifics:
	1) librdmacm is used for connection management
	2) Transport layer is implemented over IB RC QP
	3) Short messages handling is implemented as an eager protocol. It uses signaled RDMA write with imm data ibv_post_send. Preregistered memory region is splitted into N buffers in a round robin manner. Every buffer has a service data with a status field: free or busy. Busy means that the message has not been processed yet and the buffer can't be reused). Send buffer is released by the receiver after processing of WC extracted from RCQ (ibv_poll_cq) via an update of the status field on the sender side by RDMA write.
	4) large messages handling is implemented as a rndv read protocol with registration/deregistration of user's memory region every time the massage being sent.
	5) Memory management on a critical path is controlled by memory pools. They have some amount of pre-allocated memory and perform extra allocations in case of insufficient amount of resources.

Design specifics:
	1) Internal request object, which is an entry of OFI completion queues and has its own complex state (eager + rndv)
	2) For eager messaging "rndv state" is always set to "not used"
	3) The rndv protocol flow is controlled by a set of special eager messages: rts, ack. The "eager state" field is applicable for the messages.
	4) States transition is performed inside of state machine handlers which are organized as a 3-d array: "eager state", "rndv state" and "event".
	5) There are several types of events: coming from upper level, lower level and internal. The upper level events are initiated by OFI send/recv calls, the lower level events are generated by incoming packets or local completions, the internal "ready to send" event is triggered when the availability  check for required resources is successful.

### CODE REVIEW

File: prov/verbs/src/verbs_checks.h [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+        return -FI_ENODATA;
+    }
+
+    switch (attr->threading) {
+    case FI_THREAD_UNSPEC:
+    case FI_THREAD_ENDPOINT:
+        break;
+    case FI_THREAD_SAFE:
+    case FI_THREAD_FID:
+    case FI_THREAD_DOMAIN:
```
Comment (ddurnov) [inline prov/verbs/src/verbs_checks.h]: Please make FI_THREAD_DOMAIN supported as well. Even if it isn't implemented we should pass the check in order to make MPICH/CH3/netmod/OFI work

**Comment by shefty** [comment]
Before merging this, I think some staging patches are needed first.  An initial set of patches that does nothing other than refactor the existing code into separate files should precede a patch that adds the RDM support.  As this commit stands, it is an entire replacement of the existing code, making it difficult to see what changes have actually occurred to the current code base.  There have been recent commits to the upstream code that are being lost here.

**Comment by shefty** [comment]
I can assist with the staging patches.  Those should just result in moving the existing code from fi_verbs.c into new files.  Just let me know.

**Comment by shefty** [comment]
At the risk of duplicating work, I will go ahead and start creating/submitting patches to factor out code from the existing provider into separate files.

**Comment by evgeny-leksikov** [comment]
No  needed, I've almost done it locally.

**Comment by evgeny-leksikov** [comment]
Done, EP_MSG status should conform to commit 2ca5b2d

**Comment by shefty** [comment]
This series isn't mergeable as is.  Each patch in the series needs to stand by itself without breaking the build or existing functionality.  I can't merge patch that is immediately followed by a patch that reverts it.  Nor can I merge a patch that breaks the build, followed by a patch to fix it that's part of the same series.  (As an FYI, I don't know if you've used quilt or stgit, but those will help keep a workable patch series where any individual patch may require updates.)

I am in the process of reshaping the changes into a new set of commits.  Some should be ready for merging by tomorrow, with a target to have most of the changes in by the end of the week.  But I'm reviewing the changes as I go, so it could take longer depending on the code structure.

**Comment by shefty** [comment]
Update: We've starting merging in changes that separate out the existing code into separate files.  I've analyzed the differences between this tree and master, making notes of areas where common functionality changed.  Those changes appear fairly limited.

I'm not done reviewing the RDM code, but I have already noted the following.

The coding style, including indentation, needs updating.  This is minor.
A new version of lists is included as part of the change.  There already exists list implementations in include/fi_list.h.  That needs to be used instead, and expanded if needed.
Hashing routines need to move to common code.  We need to ensure that we have the best implementation, as hashing will be needed for the UDP provider and PSM (I think).
There's a static inline implementation for a memory pool.  That needs to move into shared code as well, and likely out of a header file into c files.
There are new implementations of memset and memcpy.  Those need to be removed, and use the standard library provided ones, which are already optimized for the various platforms.
Most of the RDM data structures and defines should probably move into a verbs_rdm.h file, if they can't go directly into a relevant C file.

Arun and I are both working on resolving these issues, with the hope of getting this code merged by the end of next week, in time for 1.2rc1

**Comment by evgeny-leksikov** [comment]
Thanks! Let me know if I can help with something independently and do not generate new conflicts. Meanwhile, I'll replace utlist to fi_list usage and move rdm specific things to separate header file.
I have never used quilt or stgit before and any help with commit series reorganization is welcome.

**Comment by shefty** [comment]
I use stgit myself.  It's incredibly easy.  You can make due with only a handful of commands.

I have a clone of your tree locally, so as you make changes, if you just push them upstream, I can fetch them and integrate the changes into a 'cleaner' patch series.  Please let me know when you have the first set of change done.  I will move the memory pool functionality into common code.

For the hashing functions, was there a specific reason why uthash was chosen?  (I'm not questioning the choice, just wanting to know the reasoning behind it.)  I looked at the license, and it is BSD, so looks okay to integrate.

**Comment by evgeny-leksikov** [comment]
I've already replaced utlist usage with fi_list + new functionality: find_first_match and foreach (let me know if you have any objections here), but it can't be pushed right now because of bugs. I hope I'll fix it tomorrow.
WRT uthash, It was chosen before as I started working on the provider. I guess the main reason is following to utlist. Probably, fi_list was not implemented at that time at least on our internal repository.

**Comment by evgeny-leksikov** [comment]
Reworked with fi_list usage

**Comment by evgeny-leksikov** [comment]
Update: rdm/tagged specific structures and definitions are moved into verbs_rdm.h

**Comment by shefty** [comment]
Merged manually through my tree.

**Pull request closed without merging**